runtime versions of Intrinsics.reinterpret

vtjnash opened the pull request:
this implements runtime versions of all of the julia intrinsics (so that functions that use them don't need to be compiled to be run). it additionally includes a `Core.Intrinsics.intrinsic_call` invoke point for calling any of these runtime intrinsics. and it adds the name of the intrinsic to `jl_` printing.

additionally, It implements full error checking for the reinterpret/box/unbox intrinsic (fixes #12832), falling back to the the runtime version when necessary.

finally, it re-organizes some of the julia codegen architecture so that it is possible to link a version of libjulia that does not include a copy of LLVM (although I don't recommend trying, since the runtime currently can't get very far without the JIT).

~~(marked as wip simply because I'm still working on this branch, not because it couldn't be merged)~~

vtjnash committed:
WIP: start implementation of generic versions of the julia intrinsic functions

vtjnash committed:
wip: cleanup intrinsics

vtjnash committed:
introduce a call method for (runtime versions of) intrinsics. use it to have "reinterpret" do runtime error checking (fix #12832)

vtjnash committed:
create a compile option for a Julia that is not linked against llvm (other than libLLVMSupport.a for APInt support)

vtjnash committed:
improve jl_ printing for functions

vtjnash commented on src/APInt-C.cpp:
```diff
@@ -0,0 +1,486 @@
+#include "llvm-version.h"
```
need license

vtjnash commented on src/APInt-C.cpp:
```diff
@@ -0,0 +1,486 @@
+#include <llvm/ADT/APFloat.h>
+#include <llvm/Support/MathExtras.h>
+
+#define DLLEXPORT
```
???

vtjnash commented on src/APInt-C.h:
```diff
@@ -0,0 +1,78 @@
+
```
add license

vtjnash commented on src/intrinsics.cpp:
```diff
@@ -413,10 +495,7 @@ static jl_cgval_t generic_box(jl_value_t *targ, jl_value_t *x, jl_codectx_t *ctx
-                // since codegen has no idea what size it might have
-                jl_error("codegen: failed during evaluation of a call to reinterpret");
-                return jl_cgval_t();
+                return jl_cgval_t(); // TODO: XXX
```
revert this blob

tkelman commented on src/runtime_intrinsics.c:
```diff
@@ -0,0 +1,930 @@
+
+
+static inline unsigned int next_power_of_two(unsigned int val) {
+  /* this function taken from libuv src/unix/core.c */
```
license here too

vtjnash replied:
I wouldn't consider this a "substantial portion" of libuv, so i don't think it needs one (https://github.com/libuv/libuv/blob/26917ddf7a5279c5d913c12121fe955ba4de91c0/LICENSE#L14-L15)

vtjnash committed:
cleanup runtime_intrinsics DLLEXPORT and license

vtjnash committed:
simplify more functions signatures that use Intrinsics (for compile-all mode)

vtjnash committed:
avoid running and clearing the __init__ list when building output

tkelman commented on src/intrinsics.cpp:
```diff
@@ -819,43 +901,87 @@ static Value *emit_smod(Value *x, Value *den, jl_codectx_t *ctx)
     case llvmcall: return emit_llvmcall(args, nargs, ctx);
-
-    HANDLE(pointerref,2)
+#if 0
```
what's this about?

vtjnash replied:
you can flip this switch and codegen will only use runtime intrinsics, never inline / optimized :O

it's a useful mode for showing that the runtime versions pass all of the julia tests, but its rather slow (it ends up doing too much re-boxing due to the current codegen design).

tkelman replied:
I see, that makes sense. Is this worthy of a named option define? and/or an explanatory comment...

tkelman commented on src/runtime_intrinsics.c:
```diff
@@ -0,0 +1,930 @@
+
+static inline char signbitbyte(void *a, unsigned bytes) {
+    // sign bit of an signed number of n bytes, as a byte
+    return signbit(((signed char*)a)[bytes-1]) ? ~0 : 0;
```
msvc errors on this
```
d:/code/msys64/home/Tony/julia/src/runtime_intrinsics.c(89) : error C2668: 'sign
bit' : ambiguous call to overloaded function
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\math.h(32
9): could be 'bool signbit(long double)'
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\math.h(32
4): or       'bool signbit(double)'
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\math.h(31
9): or       'bool signbit(float)'
        while trying to match the argument list '(signed char)'
```

tkelman commented on src/runtime_intrinsics.c:
```diff
@@ -0,0 +1,930 @@
+    typedef intrinsic##_t select_##intrinsic##_t[6]; \
+    static inline intrinsic##_t select_##intrinsic(unsigned sz, select_##intrinsic##_t list) \
+    { \
+        return list[select_by_size(sz)] ?: list[0]; \
```
gnu extension

tkelman commented on src/runtime_intrinsics.c:
```diff
@@ -0,0 +1,930 @@
+        /* TODO: this memcpy assumes little-endian,
+         * for big-endian, need to align the copy to the other end */ \
+        memcpy(pa2, pa, isize);
+        memset(pa2 + isize, getsign(pa, isize), osize2 - isize);
```
`d:/code/msys64/home/Tony/julia/src/runtime_intrinsics.c(292) : error C2036: 'void *' : unknown size`

tkelman commented on src/runtime_intrinsics.c:
```diff
@@ -0,0 +1,930 @@
+// checked arithmetic
+#define check_sadd(a,b) \
+        /* this test is a reduction of (b > 0) ? (a + b >= typemin(a)) : (a + b < typemin(a)) ==> overflow */ \
+        (b > 0) == (a >= (((typeof(a))1) << (8 * sizeof(a) - 1)) - b)
```
```
d:/code/msys64/home/Tony/julia/src/runtime_intrinsics.c(842) : error C2143: synt
ax error : missing ')' before 'constant'
d:/code/msys64/home/Tony/julia/src/runtime_intrinsics.c(842) : error C2059: synt
ax error : ')'
d:/code/msys64/home/Tony/julia/src/runtime_intrinsics.c(842) : error C3861: 'typ
eof': identifier not found
```

vtjnash committed:
keep TypeVar out of typeinf results, and add assertion

vtjnash committed:
store a minimal list for jl_module_init_order

reduces runtime lookup effort. but the real reason is that it is a workaround because of the replacing of Core.Inference

vtjnash committed:
remove gcc-isms from runtime-intrinsics code

tkelman commented:
okay if I commit https://gist.github.com/6c318f37be5a77b42ca7 ? MSVC is strange.

JeffBezanson commented on src/anticodegen.c:
```diff
@@ -0,0 +1,50 @@
+jl_value_t *jl_static_eval(jl_value_t *ex, void *ctx_, jl_module_t *mod,
+                           jl_value_t *sp, jl_expr_t *ast, int sparams, int allow_alloc)
+{
+    return NULL;
```
Seems like it should be possible to always include a useful version of this function.

JeffBezanson commented on src/intrinsics.h:
```diff
@@ -0,0 +1,207 @@
+#endif
+JL_CALLABLE(jl_f_intrinsic_call)
+{
+    JL_NARGSV(intrinsic_call, 1);
```
I'd rather have this function in a .c file.

JeffBezanson commented:
Cool!

It looks like `intrinsic_call` is not used yet? Is the idea to use this at all intrinsic call sites eventually?

vtjnash commented:
Long term, yes. Or make them all Function with this as the fptr and the id in the env (no linfo). It's all a bit dependent on what happens in the jb/functions branch
